src/query_engine/AtomDB.cc

```diff
@@ -140,16 +141,50 @@ void AtomDB::mongodb_setup() {
     }
 }
 
-shared_ptr<atomdb_api_types::HandleList> AtomDB::query_for_pattern(shared_ptr<char> pattern_handle) {
-    redisReply *reply = (redisReply *) redisCommand(this->redis_single, "SMEMBERS %s:%s", REDIS_PATTERNS_PREFIX.c_str(), pattern_handle.get());
-    if (reply == NULL) {
-        Utils::error("Redis error");
-    }
-    if (reply->type != REDIS_REPLY_SET && reply->type != REDIS_REPLY_ARRAY) {
-        Utils::error("Invalid Redis response: " + std::to_string(reply->type));
+vector<string> AtomDB::query_for_pattern(std::shared_ptr<char> pattern_handle) {
+    unsigned int redis_cursor = 0;
+    bool redis_has_more = true;
+    string command;
+    redisReply *reply;
+
+    vector<string> results;
+
+    while (redis_has_more) {
+        command = (
+            "ZRANGE "
+            + REDIS_PATTERNS_PREFIX
+            + ":"
+            + pattern_handle.get()
+            + " "
+            + to_string(redis_cursor)
+            + " "
+            + to_string(redis_cursor + REDIS_CHUNK_SIZE - 1)
+        );
+
+        reply = (redisReply *) redisCommand(this->redis_single, command.c_str());
+
+        if (reply == NULL) {
+            Utils::error("Redis error");
+            break;
+        }
+
+        if (reply->type != REDIS_REPLY_SET && reply->type != REDIS_REPLY_ARRAY) {
+            Utils::error("Invalid Redis response: " + std::to_string(reply->type));
+            freeReplyObject(reply);
```

Comment:
`reply` must be freed before calling `Utils::error` b/c it throws an exception:
```suggestion
			auto reply_type = std::to_string(reply->type);
            freeReplyObject(reply);
            Utils::error("Invalid Redis response: " + reply_type);
```